assets/src/edit-story/components/fontPicker/scrollList.js

```diff
@@ -0,0 +1,134 @@
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { useEffect, useCallback, useRef } from 'react';
+import styled from 'styled-components';
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+import debounce from '../../utils/debounce';
+
+const List = styled.ul.attrs({ role: 'listbox' })`
+  width: 100%;
+  overflow-x: hidden;
+  overflow-y: auto;
+  overscroll-behavior: none auto;
+`;
+
+const Item = styled.li.attrs({
+  tabIndex: '0',
+  role: 'option',
+})`
+  overflow: hidden;
+`;
+
+function ScrollList({
+  items,
+  itemRenderer,
+  onScroll,
+  currentOffset,
+  className,
+}) {
+  const ref = useRef();
+  const itemOffsets = useRef([]);
+  const numItems = items.length;
+
+  const setOffset = useCallback(
+    (index) => (node) => {
+      if (!node) {
+        return;
+      }
+      itemOffsets.current[index] = node.offsetTop;
+    },
+    []
+  );
+
+  useEffect(() => {
+    const node = ref.current;
+    if (!node) {
+      return undefined;
+    }
+
+    const handleScroll = debounce((evt) => {
+      if (!evt.target) {
+        return;
+      }
+      const {
+        target: { scrollTop, clientHeight },
+      } = evt;
+      const firstIndexInView = itemOffsets.current.findIndex(
+        (y) => y > scrollTop
+      );
+      const firstIndexOutOfView = itemOffsets.current.findIndex(
+        (y) => y > scrollTop + clientHeight
+      );
+      const lastIndexInView =
+        firstIndexOutOfView === -1 ? numItems : firstIndexOutOfView - 1;
+      onScroll(firstIndexInView, lastIndexInView);
+    }, 100);
+
+    // Invoke now
+    handleScroll({ target: node });
+    // And when scroll changes (but debounced)
+    node.addEventListener('scroll', handleScroll);
+    return () => node.removeEventListener('scroll', handleScroll);
+  }, [onScroll, numItems]);
+
+  // If current offset changes, scroll to that one
+  useEffect(() => {
+    const node = ref.current;
+    if (currentOffset === -1 || !node) {
+      return;
+    }
+    const offset = itemOffsets.current[currentOffset];
+    node.scrollTo(0, -(offset + node.clientHeight / 2));
+  }, [currentOffset]);
+
+  // Trim to current length of list
+  useEffect(() => {
+    itemOffsets.current = itemOffsets.current.slice(0, numItems);
+  }, [numItems]);
+
+  return (
+    <List
+      className={className}
+      aria-multiselectable={false}
+      aria-required={false}
+      ref={ref}
+    >
+      {items.map((item, index) => (
+        <Item ref={setOffset(index)} key={index}>
```

Comment:
That's a lot of measurements and possibly interspersed with renders. Do we need offsets? Aren't they all the same height?

Reply:
For now, yes - but in the future search and previously used fonts will add some dividers changing this. I've changed it a bit now to just be a list of node references to be able to focus them (better focus management coming later).